File: backend/utils/translateannouncements.py
import json
from typing import List, TypedDict
from ai import GroqClient
from prompts.translatetranslateannouncements import GetPrompt

from typing import Optional

class Announcement(TypedDict, total=False): 
    title: str
    link: str
    id: Optional[str]


async def translate_announcements(
    announcements: List[Announcement], target_language: str
) -> List[dict]:
    try:
        if not announcements:
            return []

        prompt = GetPrompt(announcements, target_language)

        response = GroqClient.chat.completions.create(
            model="openai/gpt-oss-20b",
            messages=[
                {
                    "role": "system",
                    "content": f"You are an expert translator for Indian government documents. "
                               f"Translate accurately into {target_language} while maintaining official tone and terminology. "
                               f"Respond ONLY in valid JSON format."
                },
                {
                    "role": "user",
                    "content": (
                        f"{prompt}\n\nReturn output strictly as a JSON object with this structure:\n"
                        "{ \"translations\": [ {\"title\": \"...\", \"link\": \"...\"} ] }"
                    )
                }
            ],
            response_format={"type": "json_object"}
        )

        content = getattr(response.choices[0].message, "content", None)
        if not content:
            print("Translation API returned empty content")
            return []

        try:
            data = json.loads(content)
        except json.JSONDecodeError as e:
            print(f"JSON parsing failed: {e}, content: {content}")
            return []

        translated = data.get("translations", [])

      
        results = []
        for original, trans in zip(announcements, translated):
            results.append({
                "original_title": original.get("title"),  
                "title": trans.get("title"),           
                "link": trans.get("link"),
                "id": original.get("id") if "id" in original else None
            })

        return results

    except Exception as e:
        print(f"Translation error: {e}")
        return []

File: backend/src/utils/translateannouncements.ts
import { groq } from "../services/Groq"
import { Get_Prompt } from "../prompts/translateannouncements"
import { ObjectId } from 'mongodb';

interface Announcement {
    title: string;
    source: string;
    _id: ObjectId,
    type: string
    content: string,
    summary: string
}

export interface TranslatedAnnouncement {
    title: string;
    source: string;
    _id: ObjectId;
    type: string
    content: string
    summary: string
}

export const translateannouncements = async (
    announcements: Announcement[],
    target_lan: string
): Promise<TranslatedAnnouncement[]> => {

    const prompt = Get_Prompt(announcements, target_lan);

    try {

        const response = await groq.chat.completions.create({
            messages: [{ role: "system", content: prompt }],
            model: "openai/gpt-oss-20b",
            response_format: { type: "json_object" }
        });

        const text = response.choices?.[0]?.message?.content;

        if (!text) {
            throw new Error("No response content received from API");
        }


        const translatedData = JSON.parse(text);


        if (!translatedData.announcements || !Array.isArray(translatedData.announcements)) {
            throw new Error("Invalid response format: expected announcements array");
        }


        const result: TranslatedAnnouncement[] = translatedData.announcements.map((item: any, index: number) => {
            if (!item.title || !item.link) {
                console.warn(`Missing title or link for announcement at index ${index}`);
            }

            return {
                title: item.title || announcements[index]?.title || "Untitled",
                link: item.link || announcements[index]?.source || "",
                _id: item._id || announcements[index]?._id || "",
                type: item.type || announcements[index]?.type || "",
                summary: item.content || announcements[index]?.content || ""
            };
        });

        return result;

    } catch (error) {
        console.error("Translation error:", error);

        return announcements
    }
}

export default translateannouncements;